External drag and drop support for compose application

Walingar opened the pull request:
## Proposed Changes

Introduce drag and drop support from external resources to compose application.
For now, for desktop only.

## Testing

It can be tested by introduced `androidx.compose.ui.dnd.ExternalDragTest` or manually by `androidx.compose.desktop.examples.dnd.Main_jvmKt#main`

## Issues Fixed

Fixes: https://github.com/JetBrains/compose-jb/issues/222

Walingar committed:
Introduce external file drag and drop support for desktop

Walingar committed:
Introduce more drop data types for external drag support

Walingar committed:
Add docs to external drag API

Walingar committed:
Add offset to dragStart callback like in onDrag function

Walingar committed:
Move awt point -> offset logic closer to drag event

Walingar committed:
Introduce tests for external drag support

Walingar commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,419 @@
+    }
+}
+
+// @VisibleForTesting
```
Not sure if we can add annotations library to this gradle module. I've seen them only for android target

dima-avdeev-jb replied:
VisibleForTesting needs for public functions only to share between modules.
In your case, it's redundant.

Walingar replied:
Thanks! I often use it for classes/functions that can be private, but I have to provide internal visibility for tests. But no one in the module should use it except tests
If you think it is redundant, I'll remove it

dima-avdeev-jb replied:
Yeah, also this is a better description of VisibleForTesting in doc comment:

```
/**
 * Denotes that the class, method or field has its visibility relaxed, so that it is more widely
 * visible than otherwise necessary to make code testable.
 * <p>
 * You can optionally specify what the visibility <b>should</b> have been if not for
 * testing; this allows tools to catch unintended access from within production
 * code.
 * <p>
 * Example:
 * <pre><code>
 *  &#64;VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
 *  public String printDiagnostics() { ... }
 * </code></pre>
 *
 * If not specified, the intended visibility is assumed to be private.
 */
 public @interface VisibleForTesting {
```

Walingar replied:
It means that it is better to provide this annotation)

dima-avdeev-jb replied:
Sorry, maybe I am wrong

Walingar committed:
Mark external drop API as experimental

Walingar commented:
Made few force-pushes with rebasing on up-to-date main branch and removed links to the issues in commit messages, since it spams there :(

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+ */
+@ExperimentalComposeUiApi
+@Composable
+fun Modifier.onExternalDrag(
```
Maybe call it `onExternalDragAndDrop`, to clarify the context?

Walingar replied:
I used such naming similar to `androidx.compose.foundation.gestures.DragGesture_skikoKt.onDrag`

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+    val window = LocalWindow.current ?: return@composed Modifier
+
+    var componentCoordinates by remember { mutableStateOf<LayoutCoordinates?>(null) }
+    val boundsInWindow = componentCoordinates?.boundsInWindow()
```
Maybe it makes sense for this to be a `derivedState`?

Walingar replied:
How can we use it here? As you can see we modify componentCoordinates in `Modifier.onGloballyPositioned` later, so when global position changes boundsInWindow will be recalculated. Since derivedState provides nonMutable state that depends on `State` inside it

Walingar replied:
I decided to use only bounds (Rect) instead of LayoutCoordinates, will publish it later

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+
+    // components drag handlers depend on the component position,
+    // so we should reinstall handlers with a new component position when it was changed
+    DisposableEffect(window, boundsInWindow) {
```
Does the `DisposableEffect` have to take `boundsInWindow` as a key? If the drop-area is animating its size/position, the drag will be restarting all the time.
Is there's a way to arrange for the drag to not restart in this case, but just update the target area? Maybe with `rememberUpdatedState`?

Walingar replied:
Thank you! This is a really good point! I'll write a test for that and see what we can do

Walingar replied:
Will rewrote it, so boundsInWindow will be passed to Swing in `onGloballyPositioned`

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+ * @param onDragStart will be called when the pointer with external content entered the component.
+ * @param onDrag will be called for all drag events inside the component.
+ * @param onDrop is called when the pointer is released with [DropData] the pointer held.
+ * @param onDragCancel is called if the pointer exited the component bounds or unknown data was dropped.
```
Style guide says we're supposed to fit into 100 columns.

Walingar replied:
Hm, weird that IDEA doesn't move it to the new line. Where do you see this into styleguide?

m-sasha replied:
Not sure; @igordmn told me.

igordmn replied:
This was worked perfectly in Android Studio in AOSP, but we have bugs in our fork, and this limit doesn't always apply.

There is an idea to increase it for our fork. It will complicate upstreaming into AOSP, but that is another discussion.

Walingar replied:
Should I manually controll this length and change it here or it is ok?

igordmn replied:
Check your settings in IDEA, if you will find that you don't use `Project`, switch to it, and then comply to it:
![image](https://user-images.githubusercontent.com/5963351/224427268-61a0223b-48bd-4897-b5cc-224f63ca4cb8.png)

Walingar replied:
I use the project one, weird

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+        },
+        // drag moved inside window, we should calculate whether drag entered/exited components or just moved inside them
+        onDragInsideWindow = { newWindowDragCoordinates ->
+            for ((_, handler) in handlers) {
```
Maybe 
```
for (handler in handlers.values) {
```
instead?

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+        },
+        // notify components on window border drag exited window
+        onDragExit = {
+            for ((_, handler) in handlers) {
```
`handlers.values` here too, maybe?

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+    ): Int {
+        isActive = true
+
+        handlers[idsCounter] = ComponentDragHandler(componentCoordinates, onDragStart, onDrag, onDragCancel, onDrop)
```
Maybe update `idsCounter` immediately, just to be extra safe to avoid concurrency/reentrance issues.

Walingar replied:
Everything here goes in one thread (AWT), because of that idsCounter is not atomic

m-sasha replied:
Yes, `idsCounter` itself doesn't need to be atomic, but it's safer if the update is atomic in the thread, like 
```
val nextId = idsCounter++
handlers[nextId] = ...
...
return nextId
```
because who knows what onDragStart will do. It could for whatever reason call `installComponentDragHandler` itself.

It's really unlikely, so just to be on the really safe side.

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+
+    // takes title bar and other insets into account
+    private fun Point.windowOffset(): Offset {
+        val offsetX = (x - window.insets.left) * density
```
I don't know whether it matters, but `window.density.density` is `window.graphicsConfiguration.defaultTransform.scaleX`. Maybe you'd want to use `scaleY` for the top insets.

Walingar replied:
This is a good point! Firstly, I used scaleY here, but then I looked at other places and saw that window.density.density is used everywhere, so I think we should be consistent and if user uses different scales for X and Y (which is weird?) we have to fix it everywhere at once. What do you think?

igordmn replied:
We usually don't distinguish between scaleX/scaleY, as there is no use case for that in Compose. We retrieve the screen density from this property because there is no other way.

Walingar replied:
So, I'll keep it, right?

m-sasha reviewed (approved):
Mostly nitpicks. A great PR overall.

Walingar committed:
Use handlers.values

igordmn commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+    /**
+     * Represents an image drag and dropped to an application.
+     */
+    data class Image(val painter: Painter) : DropData
```
1. We lack API for a generic byte array.
2. We probably should provide API that don't read all data straight away. I.e. users might want to check the mimeType first, change UI element (or the cursor) if it supported this type, and read data only when user unpressed the mouse.

What do you think, if we implement `DropData` this way?

```
interface DropData {
  val mimeType: String
  fun readBytes(): ByteArray

  interface FileList : DropData {
     fun readFileList(): List<String>
  }

  class Image : DropData {
    fun readImage(): Painter
  }

  interface Text : DropData {
     fun readText(): String
  }
}
```

mimeType we can retrieve from `getTransferDataFlavors` (`The array should be ordered according to preference for providing the data (from most richly descriptive to least descriptive)`.

`readBytes()`, I hope is supported by every mime type, when we pass `application/octet-stream` to `getTransferData`

Walingar replied:
This sounds good! But I doubt about mimeType providing, since it can be "image/x-java-image" is it ok?
I am thinking that we can provide a list of mimeTypes like java does

igordmn replied:
>  since it can be "image/x-java-image" is it ok?

Not very pretty and crossplatform (when we port it to other platforms), but If we won't find another way, it can still be useful. People can check individual mimeType's on each platform for different use cases that needed for their application, and handle them all.

In the future, when we implement other platforms, we can do research, and change mimeType to something neutral (I think we need to keep this API experimental until we do that)

Walingar replied:
As I checked there is no `application/octet-stream` for every data, at least it won't be clear what the data is for file. Content? File name? etc?, So I would keep mimeType for all, so that user can check it and provide basic interfaces like Image, Text, Uri, FileList

igordmn commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+     *
+     * @param mimeType mimeType of the [content] such as "text/plain", "text/html", etc.
+     */
+    data class Text(val content: String, val mimeType: String?) : DropData
```
Is it possible to have null? Can we pass "text/plain" if we have null? If we can't, better to describe in the docs when this can happen.

igordmn commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+    /**
+     * Represents list of files drag and dropped to an application in a raw [java.net.URI] format.
+     */
+    data class FilesList(val rawUris: List<String>) : DropData
```
Unfortunately we can't use sealed/data in public API's.
- sealed classes is hard to extend without breaking backward compatibility
- data classes is also hard to extend (https://jakewharton.com/public-api-challenges-in-kotlin/)

Because of that, Google has policies in Compose:

- implement copy/hashCode,equals manually (or via IDEA tooling)
- instead of sealed class/interface - use just class/interface

Walingar committed:
Use componentBounds instead of LayoutCoordinates for external drag support

igordmn commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+@Composable
+fun Modifier.onExternalDrag(
+    enabled: Boolean = true,
+    onDragStart: (Offset) -> Unit = {},
```
Can we extend API, so we can know what mime type is dragging right now? I.e. we can pass `DropData` here, if we change its interface as described above.

Walingar replied:
Not sure, will take a look at it and let you know

Walingar replied:
Checked, we can, will do so

Walingar committed:
Introduce test to check that drag enter/cancel are not spammed on component bounds change

igordmn reviewed (commented):
Thanks!

I reviewed only API and behavior. Let's m-sasha do a full review.

igordmn commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+ * limitations under the License.
+ */
+
+package androidx.compose.ui.dnd
```
Better to move it to `androidx.compose.ui` (we won't have much functionality for Drag&Drop)

Walingar replied:
What's the difference? What if I want to extract some classes from this file? I am not sure that I want to waste `androidx.compose.ui` Or are any reasons to put it there?

igordmn replied:
We enter into the very hard scientific problem how to name things :).

But if we just look at the other packages in the project (especially in foundation and material), we see that we create individual packages only if there will be a lot of functionality in it, not just one functionaility (no matter how it be decomposed by classes or functions). There are some exceptions, when people created small packages though.

Drag&drop seems a minor feature, and it seems there is no hypothetical functionality that can be put in this package as well in the future?

Walingar replied:
Thank you, got it! Will move it out of dnd package!

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,424 @@
+    onDragStart: (Offset) -> Unit = {},
+    onDrag: (Offset) -> Unit = {},
+    onDragCancel: () -> Unit = {},
+    onDrop: (DropData) -> Unit = {},
```
These need to be `rememberUpdatedState`, or you'll call the wrong functions if they change.

Walingar committed:
Wrap lambdas in rememberUpdatedState

Walingar committed:
Add experimental compose api opt in

Walingar committed:
Don't reinstall component drag handlers on recomposition, just provide new component bounds to Swing

Walingar committed:
Extract common code for drag event handling

Walingar committed:
Update idsCounter immediately and use current handleId

Walingar committed:
Change doc

Walingar committed:
Update tests: don't notify onDrag if onDragEntered notified

Walingar commented:
@m-sasha thank you for your review! I've fixed most of your comments, most importantly, I've rewritten the code that handles component bounds changes, please take a look at it.
Igor send review about API, I'll handle it a bit later, it won't affect implementation a lot (only API + AwtWindowDragTargetListener)

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,459 @@
+            val bounds = componentBoundsHolder[handleId] ?: continue
+            action(handler, bounds)
+        }
+    }
```
Nice! But maybe `forEachPositionedComponent` instead.

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,459 @@
+    }
+    val window = LocalWindow.current ?: return@composed Modifier
+
+    val onDragStartState by rememberUpdatedState(onDragStart)
+    val onDragState by rememberUpdatedState(onDrag)
+    val onDragCancelState by rememberUpdatedState(onDragCancel)
+    val onDropState by rememberUpdatedState(onDrop)
```
That doesn't achieve the desired goal, because `installComponentDragHandler` stores the initial values of the functions and never updates them. You can fix it (and it even looks better, I think) like this: 
```
    val componentDragHandler = rememberUpdatedState(
        AwtWindowDropTarget.ComponentDragHandler(
            onDragStart = onDragStart,
            onDrag = onDrag,
            onDragCancel = onDragCancel,
            onDrop = onDrop
        )
    )
   ...
    private val handlers = mutableMapOf<Int, State<ComponentDragHandler>>()
   ...
    fun installComponentDragHandler(
        componentDragHandlerState: State<ComponentDragHandler>
    )
   ...
```

Walingar replied:
Ah, you are right! Thank you! Will fix it

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,459 @@
+            windowDragCoordinates = newWindowDragCoordinates
+            forAllPositionedComponents { handler, componentBounds ->
+                handleDragEvent(
+                    handler,
+                    oldComponentBounds = componentBounds, currentComponentBounds = componentBounds,
+                    oldDragCoordinates, newWindowDragCoordinates
```
Not sure about the style-guide, but I personally prefer to either put all arguments on one line, or one argument per line. @igordmn can say definitively whether it's important.

m-sasha replied:
Hmm, the [Kotlin style-guide](https://kotlinlang.org/docs/coding-conventions.html#method-calls) says to put related arguments on the same line, so maybe I'm wrong here.

Walingar replied:
Yes, I did accroding to kotlin style-guide

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,459 @@
+    }
+
+    Modifier
+        .onGloballyPositioned { position ->
+            // provide new component bounds to Swing to properly detect drag events
+            val dropTarget = window.dropTarget as? AwtWindowDropTarget
+                ?: return@onGloballyPositioned
+            val handleIdToUpdate = componentDragHandleId ?: return@onGloballyPositioned
+            val componentBounds = position.boundsInWindow()
+            dropTarget.updateComponentBounds(handleIdToUpdate, componentBounds)
+        }
```
Great!

m-sasha commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/dnd/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,459 @@
+     * @param handleId id provided by [installComponentDragHandler] function
+     */
+    fun stopDragHandling(handleId: Int) {
+        val handler = handlers[handleId]
+        val componentBounds = componentBoundsHolder[handleId]
```
You could do the `remove` here immediately.

Walingar committed:
Remove VisibleForTesting, can be added in the future if needed

Walingar committed:
Move external drag support to androidx.compose.ui package

Walingar committed:
Rename forAll -> forEach

Walingar committed:
Merge value acquiring and remove

Walingar committed:
Take changed callbacks into account in AwtWindowDropTarget

Walingar commented:
@m-sasha thank you for your comments, fixed most of them (except code style, since don't get what to do better)
Also, I continue investigation of better DropData API regarding Igor's comments, looking at other UI frameworks as well.

Walingar committed:
Modify DropData API

* don't use sealed interface and data classes
* provide mimeTypes for all drop data
* provide methods to read drop data content instead of having properties

Walingar committed:
Always provide offset and drop data to external drag feature

Walingar committed:
Rename onDragCancel -> onDragExit

Walingar committed:
Rename DropData -> DragData and update javadocs

Walingar commented:
@igordmn I've updated the pull request with changes in API, please take a look at it. I'd keep it experimental, since there are a few things that can be improved, but we can decide how to do that only according to the feedback.

igordmn commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,474 @@
+     *
+     * Some platform-specific mimeTypes can occur in the list. It may be changed in future versions.
+     */
+    val mimeTypes: List<String>
```
It seems if we can't read an arbitrary byte array, we don't need to know mime types?

If it is only needed to know if we dosn't support it, we can just use `else` branch when we check the type.

Walingar replied:
Hm, agree. I thought about providing a way to show UI like "Please, DnD Image" when user dragged smth else, but you are right that it can be done by checking the needed type

igordmn commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,474 @@
+ * @see onExternalDrag
+ */
+@ExperimentalComposeUiApi
+interface ExternalDragState {
```
`State` in Compose has other meaning. Something that is mutable (and has `mutableStateOf`).

Better to name it as `ExternalDragValue` (as in TextField).

Also, it seems we can use `class` instead of `interface`?

Walingar replied:
I'd prefer to keep `interface` in the API if it is not required to create a state on the user side

igordmn replied:
> I'd prefer to keep interface in the API if it is not required to create a state on the user side

That would be inconsistent with the rest of Compose API. If something is immutable, it is a class (or "data" class). Besides, users might want to extend this API (and create on the user side)

Walingar committed:
Don't provide mimeTypes, it is not needed

Walingar committed:
Rename State -> Value

Walingar committed:
Convert interface to class for current drag state

Walingar committed:
Add immutable annotation

igordmn commented on compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/ExternalDrag.desktop.kt:
```diff
@@ -0,0 +1,462 @@
+ */
+@ExperimentalComposeUiApi
+@Immutable
+class ExternalDragValue internal constructor(
```
```suggestion
class ExternalDragValue(
```

If the thing is simple (as here), better to make it public. That makes Compose more easy to extend. If this case - without exposing the constructor, we can't write a wrapper around `onExternalDrag`.

The general rule is to expose everything that is well designed and minimal for maximum use cases.

igordmn reviewed (approved):
Thanks! API looks good now.

Left one small comment, and we good to merge

Walingar committed:
Update compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/ExternalDrag.desktop.kt

Co-authored-by: Igor Demin <igordmn@users.noreply.github.com>
